Webbrowser: Changed way of synchronizing/setting properties in javasc…

ResearchDaniel opened the pull request:
…ript. Instead of adding properties to the web processor one can now set them using their path from javascript. See changelog.md and web_property_sync.html

ResearchDaniel committed:
Webbrowser: Changed way of synchronizing/setting properties in javascript. Instead of adding properties to the web processor one can now set them using their path from javascript. See changelog.md and web_property_sync.html

ResearchDaniel committed:
Webbrowser: Update MinMaxPropertyWidgetCEF according to changes. Added property.get command for it.

petersteneteg commented on modules/webbrowser/include/modules/webbrowser/properties/templatepropertywidgetcef.h:
```diff
@@ -66,12 +66,24 @@ class TemplatePropertyWidgetCEF : public PropertyWidgetCEF {
+                p->clearInitiatingWidget();
+                callback->Success("");
+            } else if (command == "property.get") {
+                if constexpr (std::is_same_v<T, bool>) {
```
we have not switched to 17 yet...

ResearchDaniel replied:
Knew this comment would come ;)
I don't think that too many people use this module and since we are switching soon anyway I will go ahead and merge it, ok?

ResearchDaniel replied:
The module is not enabled by deafault so I think it is ok

petersteneteg replied:
I was thinking we would make a clean c++14 release/tag first...

ResearchDaniel replied:
Fixed

petersteneteg reviewed (changes requested):
Looks reasonable from what I can see

ResearchDaniel committed:
Webbrowser: C++17 fix

martinfalk commented on modules/webbrowser/data/workspaces/web_property_sync.html:
```diff
@@ -2,73 +2,36 @@
-}
-
-function buttonClicked(val) { 
+<script src="../js/inviwoapi.js"></script>
```
will this work when deployed?

ResearchDaniel replied:
No, I guess we need to either put the js-files next to the workspace or inject the module path (which should be possible)

martinfalk replied:
OK. As long as the javascript files are located within the data folder of a module, they should be deployed as well. But that needs some testing.

ResearchDaniel replied:
Resolved in latest commit. Resources can now be loaded from app and module paths

martinfalk commented on modules/webbrowser/data/js/inviwoapi.js:
```diff
@@ -0,0 +1,51 @@
+		onSuccess: function(response) {
+		},
+		onFailure: function(error_code, error_message) {}
+		
```
remove empty line

martinfalk commented on modules/webbrowser/data/workspaces/web_property_sync.html:
```diff
@@ -2,73 +2,36 @@
-  <p>Text <input type="text" id="stringProperty" value="not changed" oninput="textInput(this.value)"></p>
+  <p><input type="button" value="Button" id="buttonProperty" onclick="inviwo.setProperty('Background.switchColors', {value: this.value})"></p>
+  <p>Checkbox <input type="checkbox" value="Checkbox" id="boolProperty" onchange="inviwo.setProperty('TextOverlay.enable', {value: this.checked})"></p>
+  <p>Text <input type="text" id="stringProperty" value="not changed" oninput="inviwo.setProperty('TextOverlay.text', {value: this.value})" ></p>
```
should these examples also be mentioned in the readme?

ResearchDaniel replied:
Done

martinfalk commented on modules/webbrowser/include/modules/webbrowser/properties/propertycefsynchronizer.h:
```diff
@@ -83,21 +83,29 @@ class IVW_MODULE_WEBBROWSER_API PropertyCefSynchronizer
      * Called due to cefQuery execution in message_router.html.
-     * Expects the request for be a JSON-format with key "id" of the property to change.
-     * Example {"id":"PropertyIdentifier", "value":"0.5"}
+     * Expects the request for be a JSON-format. See inviwoapi.js:
```
"the request _for be_ a JSON-format"?

Do you mean something like this?
Expects the request to be a JSON data object

martinfalk commented on modules/webbrowser/include/modules/webbrowser/properties/templatepropertywidgetcef.h:
```diff
@@ -66,12 +92,31 @@ class TemplatePropertyWidgetCEF : public PropertyWidgetCEF {
-            p->set(value);
-            p->clearInitiatingWidget();
-            callback->Success("");
+            if (command == "property.set") {
```
These available "commands" should be documented somewhere.

martinfalk replied:
What about the set functions which take more than one element?

ResearchDaniel replied:
More than one element is resolved. Documentation will have to be added.

ResearchDaniel replied:
Added documentation

martinfalk commented on modules/webbrowser/src/processors/webbrowserprocessor.cpp:
```diff
@@ -166,52 +155,6 @@ WebBrowserProcessor::WebBrowserProcessor()
     cefInteractionHandler_.setHost(browser_->GetHost());
     cefInteractionHandler_.setRenderHandler(renderHandler_);
     addInteractionHandler(&cefInteractionHandler_);
-
```
I think the option to add additional properties is quite nice. There should be at least one processor where we can add properties at runtime.

ResearchDaniel replied:
Python processor :)

ResearchDaniel replied:
This approach (setting properties directly) is more flexible, less complicated and easier to use according to me

martinfalk commented on modules/webbrowser/src/properties/propertycefsynchronizer.cpp:
```diff
@@ -78,12 +80,47 @@ bool PropertyCefSynchronizer::OnQuery(CefRefPtr<CefBrowser> browser, CefRefPtr<C
-            return (*widget)->onQuery(browser, frame, query_id, request, persistent, callback);
+        auto command = j.at("command").get<std::string>();
+        auto propCommand = std::string("property");
+        if (command == "subscribe") {
```
document?

ResearchDaniel committed:
Webbrowser: Updated CEF version to 75.0.7+g19229b6+chromium-75.0.3770.80

ResearchDaniel committed:
Webbrowser: Fix url-encoding of CEF download path

ResearchDaniel committed:
Webbrowser: Prepare for pop-up rendering

ResearchDaniel committed:
Webbrowser: Make sure that widgets are removed before properties

ResearchDaniel committed:
Webbrowser: Start on revamping the API to use callbacks on the javascript side and improving support for more properties

ResearchDaniel committed:
WebBrowser: Property->json conversion

petersteneteg commented on modules/webbrowser/include/modules/webbrowser/io/json/minmaxpropertyjsonconverter.h:
```diff
@@ -0,0 +1,112 @@
+ * \endcode
+ */
+template <typename T>
+void to_json(json& j, const MinMaxProperty<T>& p) {
```
Still need a factory here for this to work with third party proprties...

ResearchDaniel replied:
on the todo :)
Probably need a widget for each property anyway so there is one type of factory already. But sure, I’m working on it

ResearchDaniel replied:
Check out latest commit. Comments on interface?

petersteneteg replied:
Looks reasonable I think

ResearchDaniel replied:
Done

ResearchDaniel committed:
Webbrowser: Interface for Proper/JSON conversion

ResearchDaniel committed:
Webbrowser: Use JSON conversion for property synchronization.
- Added factories for conversion of properties to and from JSON
- Downgraded CEF to previously used version due to instabilities
- Updated example workspace to demonstrate property synchronization
- Added conversion for FileProperty and OptionProperty

ResearchDaniel committed:
Webbrowser: Fixed option property example

ResearchDaniel committed:
Webbrowser: Redirect https://inviwo/app and https://inviwo/modules to local app and modules directories so that scripts and other resources can be loaded when developing and after packaging.

ResearchDaniel committed:
Webbrowser: Documentation and minor fixes.

ResearchDaniel committed:
Webbrowser: Minor review fix

ResearchDaniel commented:
@petersteneteg @martinfalk This is ready for a second round review

ResearchDaniel committed:
Webbrowser merge

petersteneteg commented on modules/webbrowser/CMakeLists.txt:
```diff
@@ -79,16 +84,21 @@ set(HEADER_FILES
     include/modules/webbrowser/interaction/cefinteractionhandler.h
     include/modules/webbrowser/interaction/cefkeyboardmapping.h
-    include/modules/webbrowser/processors/propertysyncexampleprocessor.h
+	include/modules/webbrowser/io/json/boolpropertyjsonconverter.h
```
Tabs?

petersteneteg commented on modules/webbrowser/CMakeLists.txt:
```diff
@@ -126,6 +137,13 @@ set(SHADER_FILES
+set(JS_FILES
+    data/js/inviwoapi.js
+)
+ivw_group("JS Files" ${JS_FILES})
```
Since ivw_group does not know about "JS Files" I think you might want to specify the BASE arg to ivw_group

petersteneteg commented on modules/webbrowser/include/modules/webbrowser/io/json/glmjsonconverter.h:
```diff
@@ -0,0 +1,50 @@
+namespace glm {
+template <typename T, precision P, template <typename, precision> class VecType>
+void from_json(const nlohmann::json& j, VecType<T, P>& v) {
+    //for (int i = 0; i < v.length(); i++) {
```
?

ResearchDaniel replied:
Fixed

ResearchDaniel committed:
Webbrowser: Review comments and clang format

petersteneteg commented on modules/webbrowser/include/modules/webbrowser/io/json/minmaxpropertyjsonconverter.h:
```diff
@@ -0,0 +1,94 @@
+  auto end = j.count("end") > 0 ? j.at("end").get<T>() : p.getEnd();
+
+  auto rangeMin =
+      j.count("rangeMin") > 0 ? j.at("rangeMin").get<T>() : p.getRangeMin();
```
use j.value("rangeMin", p.getRangeMin()) instead.

petersteneteg replied:
or you could use 

to_json and then apply an "j.update(json overrides)" with the new values, then all the old ones automatically get updates.

ResearchDaniel replied:
Sweet

petersteneteg commented on modules/webbrowser/include/modules/webbrowser/io/json/ordinalpropertyjsonconverter.h:
```diff
@@ -27,60 +27,64 @@
-            callback->Failure(0, ex.what());
-        }
+    // Optional parameters
+    auto minVal = j.count("minValue") > 0 ? j.at("minValue").get<T>() : p.getMinValue();
```
use value

petersteneteg commented on modules/webbrowser/include/modules/webbrowser/io/json/propertyjsonconverter.h:
```diff
@@ -0,0 +1,95 @@
+/**
+ * Interface for converting between JSON and Property.
+ */
+class IVW_MODULE_WEBBROWSER_API PropertyJSONConverter {
```
Maybe the json suff should be in it's module? Guess we could also move it later if need...

ResearchDaniel replied:
no objections. We should do it quite soon after merging this PR in that case

petersteneteg commented on modules/webbrowser/include/modules/webbrowser/io/json/propertyjsonconverter.h:
```diff
@@ -0,0 +1,95 @@
+   */
+  virtual void toJSON(json& j, const Property& p) const override {
+    // Static cast will work here since we will only use the converter for its
+    // registered
```
linebreak?

petersteneteg commented on modules/webbrowser/include/modules/webbrowser/io/json/propertyjsonconverterfactory.h:
```diff
@@ -0,0 +1,51 @@
+ */
+class IVW_MODULE_WEBBROWSER_API PropertyJSONConverterFactory
+    : public StandardFactory<PropertyJSONConverter,
+                             PropertyJSONConverterFactoryObject, std::string,
```
could use "const std::string&" here

petersteneteg commented on modules/webbrowser/src/renderhandlergl.cpp:
```diff
@@ -44,7 +44,41 @@ bool RenderHandlerGL::GetViewRect(CefRefPtr<CefBrowser> browser, CefRect& rect)
+    }
+}
+
+void RenderHandlerGL::OnPopupSize(CefRefPtr<CefBrowser> browser, const CefRect& rect) {
```
Would it be possible to use this for tooltips in glfw?

ResearchDaniel replied:
Yes, it should :)

petersteneteg reviewed (approved):
Some minor suggestions, looks good from what I can see, and follow. Nice work!

ResearchDaniel committed:
Webbrowser: Review fixes

ResearchDaniel committed:
JSON/Webbrowser: Moved JSON property conversion from webbrowser to new module

petersteneteg commented on modules/json/include/modules/json/io/json/glmjsonconverter.h:
```diff
@@ -0,0 +1,50 @@
+namespace glm {
+template <typename T, precision P, template <typename, precision> class VecType>
+void from_json(const nlohmann::json& j, VecType<T, P>& v) {
+    //for (int i = 0; i < v.length(); i++) {
```
Comments?

petersteneteg commented on modules/json/include/modules/json/io/json/minmaxpropertyjsonconverter.h:
```diff
@@ -0,0 +1,92 @@
+  auto rangeMin =
+      j.value("rangeMin", p.getRangeMin());
+  auto rangeMax =
+      j.value("rangeMax", p.getRangeMax());
```
Format?

petersteneteg commented on modules/json/include/modules/json/io/json/propertyjsonconverter.h:
```diff
@@ -0,0 +1,89 @@
+   * Requires that to_json(json& j, const SrcProperty& p) has been implemented.
+   */
+  virtual void toJSON(json& j, const Property& p) const override {
+    to_json(j, static_cast<const SrcProperty&>(p));
```
Format

petersteneteg commented on modules/json/include/modules/json/io/json/templatepropertyjsonconverter.h:
```diff
@@ -0,0 +1,75 @@
+ */
+template <typename T>
+void to_json(json& j, const TemplateProperty<T>& p) {
+  j = json{{"value", p.get()}};
```
Why indent = 2

ResearchDaniel replied:
I think that the clang-format on my mac (XCode) is not behaving properly

petersteneteg reviewed (approved):
Need to format with indent = 4 in lots of files...
There is a script to run clang format on all files...

petersteneteg commented:
https://github.com/inviwo/inviwo/blob/master/tools/refactoring/format.py

ResearchDaniel committed:
Webbrowser: Warn fixes

ResearchDaniel committed:
JSON: Fixed conversion of glm types

ResearchDaniel committed:
Merge branch 'master' into feature/webbrowser-api

ResearchDaniel commented:
@petersteneteg Just a clang-format remaining for the PR

ResearchDaniel committed:
Webbrowser: Fix rendering of popups (select element)

petersteneteg committed:
Global clang format applied
